components/atom/backToTop/src/index.js

```diff
@@ -16,96 +16,82 @@ const STYLES = {
+  const [show, setShow] = useState(false)
+  const [hover, setHover] = useState(false)
+
+  const intervalRef = useRef()
+
+  useEffect(() => {
+    const container = getTarget(refContainer)
+    const handleScroll = () => {
+      const {scrollTop, scrollHeight, clientHeight} = container
+      const halfHeight = Math.floor((scrollHeight - clientHeight) / 2)
+
+      if (scrollTop > halfHeight) {
+        if (!show) {
+          setShow(true)
+        }
+      } else {
+        if (show) {
+          setShow(false)
+          setHover(false)
+        }
+      }
     }
-  }
-
-  hoverButton = e => {
-    this.setState({hover: true})
-  }
-
-  unhoverButton = e => {
-    this.setState({hover: false})
-  }
-
-  componentDidMount() {
-    this.container = getTarget(this.props.refContainer)
+    container.removeEventListener('scroll', handleScroll)
+    container.addEventListener('scroll', handleScroll)
+    return () => {
+      container.removeEventListener('scroll', handleScroll)
+    }
+  }, [refContainer, show])
 
-    const {handleScroll} = this
-    const {scrollHeight, clientHeight} = this.container
-    const halfHeight = Math.floor((scrollHeight - clientHeight) / 2)
+  const scrollStep = () => {
+    const container = getTarget(refContainer)
+    const {scrollTop} = container
+    const {current: intervalId} = intervalRef
 
-    this.container.addEventListener(
-      'scroll',
-      handleScroll.bind(this, halfHeight)
-    )
+    if (scrollTop === 0) clearInterval(intervalId)
+    if (scrollTop) container.scrollTop = scrollTop - scrollSteps
   }
 
-  componentWillUnmount() {
-    const {handleScroll, intervalId} = this
-    clearInterval(intervalId)
-    this.container.removeEventListener('scroll', handleScroll)
+  const scrollToTop = () => {
+    const intervalId = setInterval(scrollStep, scrollIntervalTime)
+    intervalRef.current = intervalId
   }
 
-  render() {
-    const {scrollToTop, hoverButton, unhoverButton} = this
-    const {iconTop: IconTop, textTop, style} = this.props
-    const {show, hover} = this.state
-    return (
-      <button
-        title="Back to top"
-        className={cx(
-          BASE_CLASS,
-          `${BASE_CLASS}--${style}`,
-          show !== null && CLASS_READY,
-          show ? CLASS_SHOW : CLASS_HIDE,
-          hover && CLASS_HOVER
-        )}
-        onMouseOver={hoverButton}
-        onMouseOut={unhoverButton}
-        onClick={scrollToTop}
-      >
-        {IconTop && (
-          <span className={CLASS_ICON}>
-            <IconTop />
-          </span>
-        )}
-        {textTop && <span className={CLASS_TEXT}>{textTop}</span>}
-      </button>
-    )
-  }
+  const hoverButton = () => setHover(true)
+  const unhoverButton = () => setHover(false)
+
+  return (
+    <button
+      title="Back to top"
+      className={cx(
+        BASE_CLASS,
+        `${BASE_CLASS}--${style}`,
+        show !== null && CLASS_READY,
+        show ? CLASS_SHOW : CLASS_HIDE,
+        hover && CLASS_HOVER
```

Comment:
Is it possible to remove hover from state and use native :hover css selector?

Reply:
https://github.com/SUI-Components/sui-components/pull/541#pullrequestreview-165479660